tests/test_connection.c

```diff
@@ -382,6 +457,109 @@ static int s_test_connection_setup_shutdown(struct aws_allocator *allocator, voi
+
+static int s_test_connection_h2_prior_knowledge(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    struct tester_options options = {
+        .alloc = allocator,
+        .no_connection = true,
+    };
+    struct tester tester;
+    ASSERT_SUCCESS(s_tester_init(&tester, &options));
+
+    /* Connect */
+    struct aws_http_client_connection_options client_options = AWS_HTTP_CLIENT_CONNECTION_OPTIONS_INIT;
+    s_client_connection_options_init_tester(&client_options, &tester);
+    client_options.prior_knowledge_http2 = true;
+    tester.client_options = client_options;
+
+    tester.server_connection_num = 0;
+    tester.client_connection_num = 0;
+    ASSERT_SUCCESS(aws_http_client_connect(&tester.client_options));
+
+    /* Wait for server & client connections to finish setup */
+    tester.wait_client_connection_num = 1;
+    tester.wait_server_connection_num = 1;
+    ASSERT_SUCCESS(s_tester_wait(&tester, s_tester_connection_setup_pred));
+
+    /* Assert that we made an http2 connection */
+    ASSERT_INT_EQUALS(tester.connection_version, AWS_HTTP_VERSION_2);
+
+    /* clean up */
+    release_all_client_connections(&tester);
+    release_all_server_connections(&tester);
+    ASSERT_SUCCESS(s_tester_wait(&tester, s_tester_connection_shutdown_pred));
+
+    ASSERT_SUCCESS(s_tester_clean_up(&tester));
+    return AWS_OP_SUCCESS;
+}
+AWS_TEST_CASE(connection_h2_prior_knowledge, s_test_connection_h2_prior_knowledge);
+
+static int s_test_connection_h2_prior_knowledge_not_work_with_tls(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+#ifdef __APPLE__ /* Something is wrong with APPLE */
+    return AWS_OP_SUCCESS;
+#endif
+    struct tester_options options = {
+        .alloc = allocator,
+        .no_connection = true,
+        .tls = true,
+        .server_alpn_list = "http/1.1",
+    };
+    struct tester tester;
+    ASSERT_SUCCESS(s_tester_init(&tester, &options));
+
+    /* Connect with prior knowledge */
+    struct aws_http_client_connection_options client_options = AWS_HTTP_CLIENT_CONNECTION_OPTIONS_INIT;
+    s_client_connection_options_init_tester(&client_options, &tester);
+    ASSERT_SUCCESS(s_tls_client_opt_tester_init(&tester, "http/1.1", aws_byte_cursor_from_c_str("localhost")));
+    client_options.tls_options = &tester.client_tls_connection_options;
+    client_options.prior_knowledge_http2 = true;
+    tester.client_options = client_options;
+
+    tester.server_connection_num = 0;
+    tester.client_connection_num = 0;
+    ASSERT_SUCCESS(aws_http_client_connect(&tester.client_options));
+
+    /* Wait for server & client connections to finish setup */
+    tester.wait_client_connection_num = 1;
+    tester.wait_server_connection_num = 1;
+    ASSERT_SUCCESS(s_tester_wait(&tester, s_tester_connection_setup_pred));
+
+    /* Assert that we made an http1.1 connection as TLS is set */
+    ASSERT_INT_EQUALS(tester.connection_version, AWS_HTTP_VERSION_1_1);
```

Comment:
Let's be explicit and just raise an error from `aws_http_client_connect()` if the user even ATTEMPTS to use prior knowledge with TLS